Add toggleable Filters Sidebar option to View Menu Bar

AbeniMatteo opened the pull request:
No substantial changes have been made from the original PR #10933.
I'll open a review for issues left to discuss.

AbeniMatteo commented on src/gui/mainwindow.cpp:
```diff
@@ -441,6 +441,7 @@ MainWindow::MainWindow(QWidget *parent)
 #endif
 
     m_propertiesWidget->readSettings();
+    m_ui->actionShowFiltersSidebar->setChecked(isFiltersSidebarVisible());
```
From https://github.com/qbittorrent/qBittorrent/pull/10933#discussion_r311366621
> Chocobo1: Have you tried it in a clean installation? I mean when there are no config file and does it still work correctly?

ArcticGems replied:
@AbeniMatteo 
Have you tested this instead? Piccirello never answered if he did it.

AbeniMatteo replied:
Yes, it works fine.

AbeniMatteo commented on src/gui/mainwindow.cpp:
```diff
@@ -1489,6 +1489,21 @@ void MainWindow::showStatusBar(bool show)
+{
+    if (show)
+        // Default sizes
+        m_splitter->setSizes({ 120, m_splitter->width() - 120 });
```
From https://github.com/qbittorrent/qBittorrent/pull/10933#discussion_r311021745
> glassez: Having no separate property for it we always restore it with default size... Isn't it a problem?

AbeniMatteo commented on src/gui/mainwindow.cpp:
```diff
@@ -1489,6 +1489,21 @@ void MainWindow::showStatusBar(bool show)
+    return sizes[0] != 0;
+}
+
+void MainWindow::showFiltersSidebar(const bool show)
```
The splitter handle should be disabled or hidden.

AbeniMatteo replied:
The handle is now disabled when the sidebar is disabled but you can still resize the sidebar to 0 width.
This needs some investigation because we can't make the left panel non collapsible.

Chocobo1 commented on src/webui/www/private/scripts/client.js:
```diff
@@ -864,6 +873,22 @@ window.addEvent('load', function() {
     });
 
+    $('showFiltersSidebarLink').addEvent('click', function(e) {
+        showFiltersSidebar = !showFiltersSidebar;
```
I would prefer you call a function that will get the sidebar state and not use a global variable.

AbeniMatteo replied:
Nice catch, it wasn't supposed to be a global variable. Check [this](https://github.com/qbittorrent/qBittorrent/pull/15409#discussion_r700970999).

AbeniMatteo replied:
Done

AbeniMatteo commented on src/webui/www/private/scripts/client.js:
```diff
@@ -292,6 +292,15 @@ window.addEvent('load', function() {
     }
 
+    // Show Filters Sidebar is enabled by default
+    let showFiltersSidebar = (LocalPreferences.get('show_filters_sidebar') === null)
```
The value is now stored in a variable and reused addressing [your issue](https://github.com/qbittorrent/qBittorrent/pull/15409#discussion_r700728813).
The function is still called multiple times but it now follows the same pattern of the other options.

Chocobo1 replied:
>The value is now stored in a variable and reused addressing your issue.

Seriously you did not address any of my comments at all... Let's try one last time:
1. Use `const showFiltersSidebar` here and create another variable in the click event handler.
2. assign `LocalPreferences.get('show_filters_sidebar')` to a local variable to avoid repetitive calls.
3. Use a helper function to get `LocalPreferences.get('show_filters_sidebar')`, that would ensure you will handle empty, invalid values consistently.

AbeniMatteo replied:
Done, let me know if there is anything else to fix.

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -527,6 +531,16 @@ void MainWindow::setTorrentAddedNotificationsEnabled(bool value)
     settings()->storeValue(KEY_NOTIFICATIONS_TORRENTADDED, value);
 }
 
+int MainWindow::getFiltersSidebarWidth() const
```
```suggestion
int MainWindow::filtersSidebarWidth() const
```

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -1489,6 +1499,44 @@ void MainWindow::showStatusBar(bool show)
+
+void MainWindow::showFiltersSidebar(const bool show)
+{
+    auto current_sidebar_width = m_splitter->sizes()[0];
```
Please check and fix coding style of entire changes.

AbeniMatteo commented:
> The handle is now disabled when the sidebar is hidden but you can still resize the sidebar to 0 width.
This needs some investigation because we can't make the left panel non collapsible.

The sidebar is now collapsible only when the user hides it. This fixes the last problem that i've found from the original PR.

Chocobo1 commented:
I don't remember this was ever mentioned: I think the changes to WebUI would be useful, but not for GUI as GUI users could already drag/collapse the sidebar so why would it need another menu item?
Currently I would propose that this PR only enhances WebUI and drop the changes to GUI.

ArcticGems commented:
> 
> 
>  but not for GUI as GUI users could already drag/collapse the sidebar so why would it need another menu item?

Because currently it looks weird (and ugly), so a function to completely hide it would be appreciated. ❤️

Chocobo1 commented:
>Because currently it looks weird (and ugly), so a function to completely hide it would be appreciated. ❤️

I don't get it, you can hide it already (without this patch) by dragging it all the way to the left.

ArcticGems commented:
> 
> 
> > Because currently it looks weird (and ugly), so a function to completely hide it would be appreciated. ❤️
> 
> I don't get it, you can hide it already (without this patch) by dragging the sidebar all the way to the left.

I mean this (default theme & black theme) =

![filters](https://user-images.githubusercontent.com/55046301/132117525-f8b23911-0a3c-44c8-a0d1-aca457a288ad.png)

Chocobo1 commented:
>I mean this (marked with red rectangle on both default theme & black theme) =

It is a completely different issue. With this PR, the layout is still the same. Try the alpha build yourself: https://github.com/qbittorrent/qBittorrent/actions/runs/1198378422

ArcticGems commented:
> 
> 
> > I mean this (marked with red rectangle on both default theme & black theme) =
> 
> It is a completely different issue. With this PR, the layout is still the same, try the alpha build yourself: https://github.com/qbittorrent/qBittorrent/actions/runs/1198378422

OK.

With this kind of feature I assumed this PR would completely hide it, why else do it.

> I think the changes to WebUI would be useful, but not for GUI as GUI users could already drag/collapse the sidebar so why would it need another menu item?
> Currently I would propose that this PR only enhances WebUI and drop the changes to GUI.

I agree with you now. 👍  Add this only to WebUI.

glassez commented:
Wouldn't this feature be useful if it completely disabled the "Filters sidebar" component, and not just hid it? Then it would save some resources by omiting to perform any filters-related processing/calculations.

ArcticGems commented:
> 
> 
> Wouldn't this feature be useful if it completely disabled the "Filters sidebar" component, and not just hid it? Then it would save some resources by omiting to perform any filters-related processing/calculations.

Makes sense. I think your suggestion interesting! 👍

AbeniMatteo commented:
The Qt part is now reimplemented and the sidebar is removed instead of being hidden when requested:
- Addressed @glassez comment about perfs
- Addressed @Chocobo1 and @ArcticGems comments about the space reserved for the handle
- Restored the ability to collpase the sidebar resizing it

Let me know if the new implementation is what you want or we should drop it and focus only on the WebUI.

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -546,7 +548,8 @@ bool MainWindow::isDownloadTrackerFavicon() const
 void MainWindow::setDownloadTrackerFavicon(bool value)
 {
-    m_transferListFiltersWidget->setDownloadTrackerFavicon(value);
+    if (m_transferListFiltersWidget != nullptr)
```
```suggestion
    if (m_transferListFiltersWidget)
```

AbeniMatteo commented:
Sorry about the delay, working on it.

Chocobo1 commented on src/webui/www/private/scripts/client.js:
```diff
@@ -87,6 +87,12 @@ const loadSelectedTracker = function() {
+const getShowFiltersSidebar = function() {
+    // Show Filters Sidebar is enabled by default
+    const show = LocalPreferences.get('show_filters_sidebar');
+    return show === null || show === 'true';
```
`return (show === null) || (show === 'true');`

AbeniMatteo replied:
Done

Chocobo1 commented on src/base/preferences.cpp:
```diff
@@ -1256,22 +1256,24 @@ void Preferences::setMainGeometry(const QByteArray &geometry)
 }
 
-void Preferences::setMainVSplitterState(const QByteArray &state)
+void Preferences::setFiltersSidebarVisible(bool value)
```
`const bool value`

AbeniMatteo replied:
Done

Chocobo1 commented on src/base/preferences.h:
```diff
@@ -340,8 +340,10 @@ class Preferences : public QObject
-    QByteArray getMainVSplitterState() const;
-    void setMainVSplitterState(const QByteArray &state);
+    bool getFiltersSidebarVisible() const;
+    void setFiltersSidebarVisible(const bool);
```
`bool value`

AbeniMatteo replied:
Done

Chocobo1 commented on src/base/preferences.h:
```diff
@@ -340,8 +340,10 @@ class Preferences : public QObject
+    bool getFiltersSidebarVisible() const;
+    void setFiltersSidebarVisible(const bool);
+    int getFiltersSidebarWidth() const;
+    void setFiltersSidebarWidth(const int width);
```
`int value`

AbeniMatteo replied:
Done

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -435,6 +428,21 @@ MainWindow::MainWindow(QWidget *parent)
 
     m_propertiesWidget->readSettings();
 
+    m_transferListFiltersWidget = nullptr;
```
Set it to `nullptr` in the header file.

AbeniMatteo replied:
Done

Chocobo1 commented on src/gui/mainwindow.cpp:
```diff
@@ -1475,6 +1483,33 @@ void MainWindow::showStatusBar(bool show)
+        m_transferListFiltersWidget = new TransferListFiltersWidget(m_splitter, m_transferListWidget, isDownloadTrackerFavicon());
+        m_splitter->insertWidget(0, m_transferListFiltersWidget);
+        m_splitter->setCollapsible(0, true);
+        m_splitter->setSizes({width, m_splitter->width() - width});
```
`(m_splitter->width() - width)`

AbeniMatteo replied:
Done

Chocobo1 reviewed (commented):
Only a few coding style issues left, it is ready for approval after fixing them.

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -1475,6 +1482,33 @@ void MainWindow::showStatusBar(bool show)
+    }
+    else
+    {
+        Q_ASSERT(m_splitter->widget(0) == m_transferListFiltersWidget);
```
I would add counterpart assertion in `if` branch or (maybe better) make the method "safe", e.g.:
```c++
if (show && !m_transferListFiltersWidget)
    //...
else if (!show && m_transferListFiltersWidget)
    //...
```

glassez commented on src/gui/mainwindow.cpp:
```diff
@@ -1801,6 +1835,14 @@ void MainWindow::on_actionShowStatusbar_triggered()
 
+void MainWindow::on_actionShowFiltersSidebar_triggered()
+{
+    const bool isVisible = static_cast<QAction *>(sender())->isChecked();
```
Using `sender()` is discouraged. `QAction:::triggered()` has `bool checked` parameter so you could use it instead.

Chocobo1 commented:
@AbeniMatteo 
Thank you!

zywo commented:
The splitter state is broken.

AbeniMatteo commented:
Can you provide some context?

xavier2k6 commented:
@AbeniMatteo @Chocobo1 @glassez @thalieht 

There seems to be an issue since this PR was introduced or maybe It's after bringing an existing issue to the forefront?!

- Add a torrent eg (gimp 2.10.30)
- It will display some trackers in the filter list
![trackers shown with filter list open](https://user-images.githubusercontent.com/42386382/171278847-8cc19fe0-862b-4bbb-a66f-413a91181664.png)
- toggle the filters sidebar to off & then on again.
- previously displayed trackers are now gone & count(s) are at `0`.
![trackers gone when filter list re-opened](https://user-images.githubusercontent.com/42386382/171279103-12da881f-0027-44b9-8ac8-02d29695224c.png)
- If I delete the torrent, this is what's shown below
![delete torrent filter status](https://user-images.githubusercontent.com/42386382/171279464-6deccf1b-f82b-4ade-a2a0-1d05aaf90b3c.png)

glassez commented:
>previously displayed trackers are now gone & count(s) are at 0.

Is this feature in master only? If so, I will fix it once #16840 is merged (to avoid making the job twice).

thalieht commented:
When splitter is moved, on qBt restart the splitter position is restored to the right of where it was left. 
If i don't move the splitter this doesn't happen.
Tested the commit before this one and it doesn't have this problem.

thalieht commented on src/gui/mainwindow.cpp:
```diff
@@ -1475,6 +1482,33 @@ void MainWindow::showStatusBar(bool show)
+        m_transferListFiltersWidget = new TransferListFiltersWidget(m_splitter, m_transferListWidget, isDownloadTrackerFavicon());
+        m_splitter->insertWidget(0, m_transferListFiltersWidget);
+        m_splitter->setCollapsible(0, true);
+        m_splitter->setSizes({width, (m_splitter->width() - width)});
```
`m_splitter->width()` in this method is 1054 for me (_after_ the insertion of `m_transferListFiltersWidget` to the splitter) while in `MainWindow::writeSplitterSettings()` it is 1904 which of course also changes `m_splitter->sizes()` used there. 
No idea where to go from there.

thalieht replied:
Update: can only be reproduced if qBt window is un-maximized and then maximized.
